### PR TITLE
coprocessor: support stream aggregation

### PR DESCRIPTION
**Pull request by zimulala**
coprocessor: support stream aggregation    
Relate to pingcap/tidb#5725

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @AndreMouche @hanfei1991 @winoros

**Comment by zimulala** [comment]
/run-integration-common-test tidb-test=pr/453 tidb-private-test=pr/453

**Comment by zimulala** [comment]
/run-integration-common-test tidb-test=pr/453 tidb=pr/5725 tidb-private-test=pr/453

**Comment by zimulala** [comment]
/run-integration-common-test tidb-test=pr/453 tidb=pr/5725 tidb-private-test=pr/453

File: src/coprocessor/dag/executor/aggregation.rs [hunk 1]
```diff
@@ -208,6 +209,181 @@ impl Executor for HashAggExecutor {
...
+impl Executor for StreamAggExecutor {
+    fn next(&mut self) -> Result<Option<Row>> {
+        if self.executed {
+            return Ok(None);
+        }
+
+        loop {
+            match self.src.next() {
+                Err(err) => return Err(err),
+
```
Comment (AndreMouche) [inline src/coprocessor/dag/executor/aggregation.rs]: redundant empty line

File: Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "tikv"
-version = "1.1.0-dev"
+version = "1.1.0-dev.1"
```
Comment (AndreMouche) [inline Cargo.toml]: Please update it in an independent PR

File: src/coprocessor/dag/executor/aggregation.rs [hunk 3]
```diff
@@ -208,6 +209,181 @@ impl Executor for HashAggExecutor {
...
+                        &self.related_cols_offset,
+                        row.handle,
+                    )?;
+                    let new_group = self.meet_new_group(&cols)?;
+                    let mut ret = if new_group {
+                        self.get_partial_result()?
+                    } else {
+                        None
+                    };
+                    for i in 0..self.agg_exprs.len() {
```
Comment (AndreMouche) [inline src/coprocessor/dag/executor/aggregation.rs]: Could we use `zip` instead here?

File: src/coprocessor/dag/executor/aggregation.rs [hunk 4]
```diff
@@ -208,6 +209,181 @@ impl Executor for HashAggExecutor {
...
+                Err(err) => return Err(err),
+
+                Ok(None) => {
+                    self.executed = true;
+                    if self.count == 1 && self.group_by_exprs.is_empty() {
+                        return Ok(None);
+                    }
+                    return self.get_partial_result();
+                }
+
```
Comment (AndreMouche) [inline src/coprocessor/dag/executor/aggregation.rs]: redundant empty line

File: src/coprocessor/dag/executor/aggregation.rs [hunk 5]
```diff
@@ -208,6 +209,181 @@ impl Executor for HashAggExecutor {
...
 
+impl Executor for StreamAggExecutor {
+    fn next(&mut self) -> Result<Option<Row>> {
+        if self.executed {
+            return Ok(None);
+        }
+
+        loop {
+            match self.src.next() {
+                Err(err) => return Err(err),
```
Comment (AndreMouche) [inline src/coprocessor/dag/executor/aggregation.rs]: use `self.src.next()?` instead?

File: src/coprocessor/dag/executor/aggregation.rs [hunk 6]
```diff
@@ -208,6 +209,181 @@ impl Executor for HashAggExecutor {
     }
 }
 
+impl Executor for StreamAggExecutor {
+    fn next(&mut self) -> Result<Option<Row>> {
+        if self.executed {
+            return Ok(None);
+        }
+
+        loop {
```
Comment (AndreMouche) [inline src/coprocessor/dag/executor/aggregation.rs]: could we use `while` to reduce `{}`?

File: src/coprocessor/dag/executor/aggregation.rs [hunk 7]
```diff
@@ -208,6 +209,181 @@ impl Executor for HashAggExecutor {
...
+            agg_func.calc(&mut agg_cols)?;
+            let agg = aggregate::build_aggr_func(self.agg_exprs[cnt].tp)?;
+            *agg_func = agg;
+        }
+
+        // group by
+        let mut group_key = Vec::with_capacity(0);
+        if !self.group_by_exprs.is_empty() {
+            let mut vals = Vec::with_capacity(self.group_by_exprs.len());
+            for d in &self.curr_group_row {
```
Comment (AndreMouche) [inline src/coprocessor/dag/executor/aggregation.rs]: It seems we can consume `self.curr_group_row` here since we redefined it on L368?

**Comment by AndreMouche** [comment]
friendly ping @hanfei1991 @winoros, Please make sure the logic is correct

**Comment by zimulala** [comment]
PTAL @AndreMouche

File: Cargo.toml (hunk 2 above)
Comment (zimulala) [reply Cargo.toml]: Because this PR needs to bind a new version that can make TiDB knows it.

File: src/coprocessor/dag/executor/aggregation.rs [hunk 8]
```diff
@@ -208,6 +209,173 @@ impl Executor for HashAggExecutor {
...
+            for (expr, func) in self.agg_exprs.iter().zip(funcs) {
+                func.update_with_expr(&self.ctx, expr, &cols)?;
+            }
+            if new_group {
+                return Ok(ret);
+            }
+        }
+        self.executed = true;
+        if self.count == 1 && self.group_by_exprs.is_empty() {
+            return Ok(None);
```
Comment (BusyJay) [inline src/coprocessor/dag/executor/aggregation.rs]: Why return `None` in this case?

File: src/coprocessor/dag/executor/aggregation.rs [hunk 9]
```diff
@@ -208,6 +209,173 @@ impl Executor for HashAggExecutor {
...
+    }
+
+    // row: &[Datum]
+    fn meet_new_group(&mut self, row: &[Datum]) -> Result<bool> {
+        if self.group_by_exprs.is_empty() {
+            return Ok(false);
+        }
+
+        let mut matched = !self.is_first_group;
+        for (cnt, expr) in self.group_by_exprs.iter().enumerate() {
```
Comment (BusyJay) [inline src/coprocessor/dag/executor/aggregation.rs]: s/cnt/i

File: src/coprocessor/dag/executor/aggregation.rs [hunk 10]
```diff
@@ -208,6 +209,173 @@ impl Executor for HashAggExecutor {
...
+            let v = box_try!(expr.eval(&self.ctx, row));
+            if matched && box_try!(v.cmp(&self.ctx, &self.next_group_row[cnt])) != Ordering::Equal {
+                matched = false;
+            }
+            self.tmp_group_row[cnt] = v;
+        }
+        let ret = !self.is_first_group;
+        if self.is_first_group {
+            self.curr_group_row = self.tmp_group_row.clone();
```
Comment (BusyJay) [inline src/coprocessor/dag/executor/aggregation.rs]: s/curr/cur

File: src/coprocessor/dag/executor/aggregation.rs [hunk 11]
```diff
@@ -208,6 +209,173 @@ impl Executor for HashAggExecutor {
...
+            self.is_first_group = false
+        }
+        if matched {
+            return Ok(false);
+        }
+        self.next_group_row = self.tmp_group_row.clone();
```
Comment (BusyJay) [inline src/coprocessor/dag/executor/aggregation.rs]: You can use `mem::swap`.

**Comment by zimulala** [comment]
Using 3 TiKV, 1 PD, 1 TiDB, the 3 TiKV are on different servers.
The table schema as follows:
```
Create Table: CREATE TABLE `lineitem` (
  `L_ORDERKEY` int(11) NOT NULL,
  `L_PARTKEY` int(11) NOT NULL,
  `L_SUPPKEY` int(11) NOT NULL,
  `L_LINENUMBER` int(11) NOT NULL,
  `L_QUANTITY` decimal(15,2) NOT NULL,
  `L_EXTENDEDPRICE` decimal(15,2) NOT NULL,
  `L_DISCOUNT` decimal(15,2) NOT NULL,
  `L_TAX` decimal(15,2) NOT NULL,
  `L_RETURNFLAG` char(1) NOT NULL,
  `L_LINESTATUS` char(1) NOT NULL,
  `L_SHIPDATE` date NOT NULL,
  `L_COMMITDATE` date NOT NULL,
  `L_RECEIPTDATE` date NOT NULL,
  `L_SHIPINSTRUCT` char(25) NOT NULL,
  `L_SHIPMODE` char(10) NOT NULL,
  `L_COMMENT` varchar(44) NOT NULL,
  PRIMARY KEY (`L_ORDERKEY`,`L_LINENUMBER`),
  CONSTRAINT `LINEITEM_FK1` FOREIGN KEY (`L_ORDERKEY`) REFERENCES `orders` (`O_ORDERKEY`),
  CONSTRAINT `LINEITEM_FK2` FOREIGN KEY (`L_PARTKEY`,`L_SUPPKEY`) REFERENCES `partsupp` (`PS_PARTKEY`,`PS_SUPPKEY`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin
```
|   | rows:6001215 && groups:1.5m && current PR | rows:6001215 && groups:1.5m && before |percentage decrease| rows:59986052 && groups:15m && current PR | rows:59986052 && groups:15m && before|percentage decrease|
|---|---|---|---|---|---|---|
|select count(L_ORDERKEY), sum(L_LINENUMBER) from lineitem group by L_ORDERKEY;| 6.43 sec  | 7.27 sec |11.6%|   31.05s | 69.76s |55.5%|
|                                                                              | 6.53 sec  | 7.36 sec |11.3% || |
|---|---|---|---|---|---|---|
|select count(L_ORDERKEY) from lineitem group by L_ORDERKEY;|4.12 sec| 4.84 sec |14.9%|19.78 |49.03 |59.7%|
|                                                           |4.22 sec| 4.61 sec |8.5% || |
|---|---|---|---|---|---|---|
|select sum(L_ORDERKEY) from lineitem group by L_ORDERKEY;| 5.00 sec |5.51 sec |9.3%|26.76 |52.55||
|                                                         | 5.15 sec  |5.72 sec| || |
|---|---|---|---|---|---|---|
|select avg(L_LINENUMBER) from lineitem group by L_ORDERKEY;|6.04 sec| 6.21 sec |2.8%|32.26 |61.08 |49.1%|
|                                                           |6.15 sec| 6.39 sec |3.8% | ||
|---|---|---|---|---|---|---|
|select avg(L_ORDERKEY) from lineitem group by L_ORDERKEY;|5.97 sec| 6.03 sec |1.0%|35.20 |60.59 |41.9%|
|                                                         |6.20 sec| 6.30 sec |1.6% | ||
|---|---|---|---|---|---|---|

**Comment by zimulala** [comment]
PTAL @winoros @AndreMouche @BusyJay

**Comment by zimulala** [comment]
PTAL @AndreMouche @BusyJay

File: src/coprocessor/dag/executor/aggregation.rs [hunk 12]
```diff
@@ -208,6 +213,178 @@ impl Executor for HashAggExecutor {
...
+
+    executed: bool,
+    group_by_exprs: Vec<Expression>,
+    agg_exprs: Vec<AggFuncExpr>,
+    agg_funcs: Vec<Box<AggrFunc>>,
+    cols: Arc<Vec<ColumnInfo>>,
+    related_cols_offset: Vec<usize>,
+    cur_group_row: Vec<Datum>,
+    next_group_row: Vec<Datum>,
+    tmp_group_row: Vec<Datum>,
```
[the remark this answers is not in the document]
Comment (zimulala) [reply src/coprocessor/dag/executor/aggregation.rs]: I want reuse `tmp_group_row`, Can I do like this? @AndreMouche  @BusyJay

File: src/coprocessor/dag/executor/aggregation.rs [hunk 13]
```diff
@@ -208,6 +213,178 @@ impl Executor for HashAggExecutor {
...
+            let v = box_try!(expr.eval(&self.ctx, row));
+            if matched && box_try!(v.cmp(&self.ctx, &self.next_group_row[i])) != Ordering::Equal {
+                matched = false;
+            }
+            self.tmp_group_row[i] = v;
+        }
+        let ret = !self.is_first_group;
+        if self.is_first_group {
+            self.cur_group_row = self.tmp_group_row.clone();
+            self.is_first_group = false
```
[the remark this answers is not in the document]
Comment (zimulala) [reply src/coprocessor/dag/executor/aggregation.rs]: Yes, I use `self.cur_group_row` to handle it in TiDB. But I think to use `is_first_group` is
 clearer, and it seems to have no side effects.

**Comment by zimulala** [comment]
PTAL @AndreMouche

**Comment by zimulala** [comment]
/run-integration-common-test tidb-test=pr/453 tidb=pr/5725 tidb-private-test=pr/453

File: src/coprocessor/dag/executor/aggregation.rs [hunk 14]
```diff
@@ -208,6 +213,177 @@ impl Executor for HashAggExecutor {
...
+            count: 0,
+        })
+    }
+
+    fn meet_new_group(&mut self, row: &[Datum]) -> Result<bool> {
+        if self.group_by_exprs.is_empty() {
+            return Ok(false);
+        }
+
+        let mut tmp_group_row = vec![Datum::Null; self.group_by_exprs.len()];
```
Comment (AndreMouche) [inline src/coprocessor/dag/executor/aggregation.rs]: ditto

File: src/coprocessor/dag/executor/aggregation.rs [hunk 15]
```diff
@@ -208,6 +213,177 @@ impl Executor for HashAggExecutor {
...
+        for (i, expr) in self.group_by_exprs.iter().enumerate() {
+            let v = box_try!(expr.eval(&self.ctx, row));
+            if matched && box_try!(v.cmp(&self.ctx, &self.cur_group_row[i])) != Ordering::Equal {
+                matched = false;
+            }
+            tmp_group_row[i] = v;
+        }
+        let ret = !self.is_first_group;
+        if self.is_first_group {
+            self.cur_group_row = tmp_group_row.clone();
```
Comment (AndreMouche) [inline src/coprocessor/dag/executor/aggregation.rs]: Why clone?

File: src/coprocessor/dag/executor/aggregation.rs [hunk 16]
```diff
@@ -208,6 +213,177 @@ impl Executor for HashAggExecutor {
...
+            *agg_func = agg;
+        }
+        // Get and decode the values of 'group by'.
+        let mut group_key = Vec::with_capacity(0);
+        if !self.group_by_exprs.is_empty() {
+            let mut vals = Vec::with_capacity(self.group_by_exprs.len());
+            for d in self.cur_group_row.drain(..) {
+                vals.push(d);
+            }
+            group_key = box_try!(datum::encode_value(vals.as_slice()));
```
Comment (AndreMouche) [inline src/coprocessor/dag/executor/aggregation.rs]: Could we use `self.cur_group_row.as_slice` directory and remove L363-L366?

File: src/coprocessor/dag/executor/aggregation.rs [hunk 17]
```diff
@@ -208,6 +213,177 @@ impl Executor for HashAggExecutor {
...
+
+    executed: bool,
+    group_by_exprs: Vec<Expression>,
+    agg_exprs: Vec<AggFuncExpr>,
+    agg_funcs: Vec<Box<AggrFunc>>,
+    cols: Arc<Vec<ColumnInfo>>,
+    related_cols_offset: Vec<usize>,
+    cur_group_row: Vec<Datum>,
+    next_group_row: Vec<Datum>,
+    is_first_group: bool,
```
[the remark this answers is not in the document]
Comment (zimulala) [reply src/coprocessor/dag/executor/aggregation.rs]: Yes, I use self.cur_group_row to handle it in TiDB. But I think to use is_first_group is clearer, and it seems to have no side effects.

**Comment by zimulala** [comment]
PTAL @AndreMouche

File: src/coprocessor/dag/executor/aggregation.rs [hunk 18]
```diff
@@ -208,6 +213,173 @@ impl Executor for HashAggExecutor {
...
+                row.handle,
+            )?;
+            let new_group = self.meet_new_group(&cols)?;
+            let mut ret = if new_group {
+                self.get_partial_result()?
+            } else {
+                None
+            };
+            let funcs = self.agg_funcs.iter_mut();
+            for (expr, func) in self.agg_exprs.iter().zip(funcs) {
```
Comment (hicqu) [inline src/coprocessor/dag/executor/aggregation.rs]: you can use `iter().zip(&mut self.agg_funcs)`.

File: src/coprocessor/dag/executor/aggregation.rs [hunk 19]
```diff
@@ -208,6 +213,173 @@ impl Executor for HashAggExecutor {
...
+        }
+        if matched {
+            return Ok(false);
+        }
+        mem::swap(&mut self.next_group_row, &mut tmp_group_row);
+        Ok(true)
+    }
+
+    // get_partial_result gets a result for the same group.
+    fn get_partial_result(&mut self) -> Result<Option<Row>> {
```
Comment (hicqu) [inline src/coprocessor/dag/executor/aggregation.rs]: Seems it never returns `None`?

**Comment by hicqu** [comment]
rest LGTM.

**Comment by zimulala** [comment]
/run-integration-common-test tidb-test=pr/453 tidb=pr/5725 tidb-private-test=pr/453

**Comment by zimulala** [comment]
/run-integration-common-test tidb-test=pr/453 tidb=pr/5725 tidb-private-test=pr/453

**Pull request merged**